src/coprocessor/cache.rs

```diff
@@ -0,0 +1,546 @@
+        // If there's no region version data in cache, get_version will
+        // return 0 and after that some thread call evict_region, we should set
+        // version to 1 to prevent cache unused data.
+        let entry = RegionDistSQLCacheEntry {
+            version: 1,
+            ..Default::default()
+        };
+        self.regions.insert(region_id, entry);
+    }
+
+    pub fn evict_region(&mut self, region_id: u64) {
+        if self.regions.contains_key(&region_id) {
+            self.evict_region_with_exist_region(region_id);
+        } else {
+            self.evict_region_with_new_region(region_id);
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    pub fn evict_region_and_enable(&mut self, region_id: u64) {
+        self.evict_region(region_id);
+        self.enable_region_cache(region_id);
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.max_size
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.map.len() == 0
+    }
+
+    pub fn update_capacity(&mut self, capacity: usize) {
+        self.max_size = capacity;
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+    }
+
+    pub fn disable_region_cache(&mut self, region_id: u64) {
+        let entry = self.regions
+            .entry(region_id)
+            .or_insert_with(Default::default);
+        entry.enable = false;
+    }
+
+    pub fn enable_region_cache(&mut self, region_id: u64) {
+        self.regions.get_mut(&region_id).map(|e| e.enable = true);
+    }
+
+    fn is_region_cache_enabled(&self, region_id: u64) -> bool {
+        self.regions.get(&region_id).map_or(true, |e| e.enable)
+    }
+
+    fn update_regions(&mut self, region_id: u64, k: DistSQLCacheKey) {
+        let opt = self.regions
+            .entry(region_id)
+            .or_insert_with(Default::default);
+        opt.cached_items.insert(k, 1);
+    }
+
+    #[inline]
+    fn remove_lru(&mut self) {
+        if let Some((_, entry)) = self.map.pop_front() {
+            self.size -= entry.size;
+        }
+    }
+}
+
+// DistSQL cache size unit is byte, default is 256MB
+pub const DEFAULT_DISTSQL_CACHE_SIZE: usize = 256 * 1024 * 1024;
+// DistSQL cache entry max size unit is byte, default is 5MB
+pub const DEFAULT_DISTSQL_CACHE_ENTRY_MAX_SIZE: usize = 5 * 1204 * 1024;
+
+pub fn new_mutex_cache(capacity: usize) -> SQLCache {
+    Mutex::new(DistSQLCache::new(capacity))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_distsql_cache() {
```

Comment:
It seems the following case may be failed according to current implementation
```
    #[test]
    fn test_distsql_cache() {
        let regionID = 10;
        let startTS1 = 1;
        let mut cache: DistSQLCache = DistSQLCache::new(200);
        let key: DistSQLCacheKey = "test1".to_string();
        let result1: Vec<u8> = vec![100, 101, 102];
        let version = cache.get_region_version(regionID);
        cache.put(regionID, key.clone(), version, result1.clone(), startTS1);
        assert_eq!(1, cache.len());
        assert_eq!(cache.get(regionID,&key,startTS1).unwrap(),&result1);

        // if data update with start_ts=9;
        cache.disable_region_cache(regionID);
        cache.evict_region_and_enable(regionID);
        // get data with start_ts = 10  cache mis
        let startTs2 = 10;
        let result2:Vec<u8>=vec![200, 201, 202]; 
        let version = cache.get_region_version(regionID);
        cache.put(regionID,key.clone(),version,result2.clone(),startTs2);
        // get data with start_ts=10,cached
        assert_eq!(cache.get(regionID,&key,startTs2).unwrap(),&result2);
        // get data with start_ts=1; cache mis
        assert!(cache.get(regionID,&key,startTS1).is_none());
        cache.put(regionID, key.clone(), version, result1.clone(), startTS1);
        // get data with start_ts=10 
        assert!(cache.get(regionID,&key,startTs2).is_none());
    }
```

Reply:
@AndreMouche I think we should only cache newest transaction's result.